kotlinx-coroutines-core/jvm/src/internal/ResizableAtomicArray.kt

```diff
@@ -28,11 +28,12 @@ internal class ResizableAtomicArray<T>(initialLength: Int) {
         val curLen = curArray.length()
         if (index < curLen) {
             curArray[index] = value
-        } else {
-            val newArray = AtomicReferenceArray<T>((index + 1).coerceAtLeast(2 * curLen))
-            for (i in 0 until curLen) newArray[i] = curArray[i]
-            newArray[index] = value
-            array = newArray // copy done
+            return
         }
+        // It would be nice to copy array in batch instead of 1 by 1, but it seems like Java has no API for that
+        val newArray = AtomicReferenceArray<T>((index + 1).coerceAtLeast(2 * curLen))
+        for (i in 0 until curLen) newArray[i] = curArray[i]
+        newArray[index] = value
+        array = newArray // copy done
```

Comment:
I'm amused by how our styles are **completely** opposite in this aspect. If I just saw the green code somewhere, I'd be very tempted to edit it, and the result would be the red code.

Just like you say when you see the opposite style by me, "minor, up to you".

Reply:
Sure, I don't mind. At one point we'll probably have to establish a common style to avoid flip-flopping, but not today.

I think I also should elaborate on my preference, though; I don't remember if I already did that, but probably not.
You try to reduce nesting, but I find nesting very useful for judging control flow at a glance.
```kotlin
if (...) {
  ...
} else {
  ...
}
...
```
If we squint, don't read the code, and just go through the control flow structures, we can still get a sense of what's going on... unless there's a `return` somewhere in those `if` statements, breaking all intuition. Such rogue `return` statements break the natural control flow that indentation demonstrates—just like `goto` or `throw`, for example, and so I reserve them for exceptional cases. I try to structure code in such a way that squinting still doesn't obscure the control flow in the common case.

For my style, the case we're discussing is cut-and-clear: the *normal* way for this function to execute is to go to the first branch. Exiting there is no emergency. If I squint, I see a resize *typically* happening, but this doesn't reflect reality.

The most semantically clear version to me is not even the red code but
```kotlin
        val curLen = curArray.length()
        if (index >= curLen) {
            val newArray = AtomicReferenceArray<T>((index + 1).coerceAtLeast(2 * curLen))
            for (i in 0 until curLen) newArray[i] = curArray[i]
            array = newArray
        }
        curArray[index] = value
```
It reflects the actual idea of "we're assigning a value, but if the index is large, we also resize".

Before introducing such an edit, I'd need to check the code to be sure that nothing relies on a happens-before induced by the volatile read and write of the array, but I can't imagine any issues, as we can pretend that this `set` is split in two operations of resizing and then assigning the new value, with resizing supposedly happening only occasionally.

Reply:
Thanks for the explanation! I'll try to internalize it for a while.

>The most semantically clear version

I second this one, it seems to do trick -- both redundant nesting is gone and the overall "intent" (what is a normal code path and what is not) is much more clear.